src/t_string.c

```diff
@@ -563,42 +564,197 @@ void mgetCommand(client *c) {
-        /* if 'NX', no need set flags SETKEY_DOESNT_EXIST. Already verified earlier! */
-        setKey(c, c->db, c->argv[j], &(c->argv[j+1]) , 0 /*flags*/);
-        incrRefCount(c->argv[j+1]);  /* refcnt not incr by setKey() */
-        notifyKeyspaceEvent(NOTIFY_STRING,"set",c->argv[j],c->db->id);
+    /* Set all key-value pairs */
+    for (int j = 0; j < kvs_count; j++) {
+        int key_idx = 1 + (j * 2);
+        int val_idx = 1 + (j * 2) + 1;
+
+        c->argv[val_idx] = tryObjectEncoding(c->argv[val_idx]);
+        setKey(c, c->db, c->argv[key_idx], &(c->argv[val_idx]), 0);
+        incrRefCount(c->argv[val_idx]);
+        notifyKeyspaceEvent(NOTIFY_STRING,"set", c->argv[key_idx], c->db->id);
     }
-    server.dirty += (c->argc-1)/2;
+
+    server.dirty += kvs_count;
     addReply(c, nx ? shared.cone : shared.ok);
 }
 
 void msetCommand(client *c) {
-    msetGenericCommand(c,0);
+    msetGenericCommand(c, 0);
 }
 
 void msetnxCommand(client *c) {
-    msetGenericCommand(c,1);
+    msetGenericCommand(c, 1);
+}
+
+void msetexCommand(client *c) {
+    int kvs_count = 0, kvs_start = -1, keys_pos = -1;
+    robj *expire = NULL;
+    int unit = UNIT_SECONDS;
+    int flags = OBJ_NO_FLAGS;
+    int expire_flag_pos = -1;
+    long long kvs_count_val;
+
+    /* Minimum arguments: MSETEX KEYS numkeys key value */
+    if (c->argc < 5) {
+        addReplyErrorArity(c);
+        return;
+    }
+    int kvs_end = -1;
+    for (int j = 1; j < c->argc; j++) {
+        char *arg = c->argv[j]->ptr;
+
+        if (!strcasecmp(arg, "KEYS")) {
+            if (keys_pos != -1) {
+                addReplyError(c, "syntax error - KEYS specified multiple times");
+                return;
+            }
+            if (j + 1 >= c->argc) {
+                addReplyError(c, "syntax error - KEYS requires numkeys argument");
+                return;
+            }
+            keys_pos = j;
+            if (getLongLongFromObject(c->argv[j + 1], &kvs_count_val) != C_OK || kvs_count_val <= 0) {
+                addReplyError(c, "invalid numkeys value");
+                return;
+            }
+            kvs_count = (int)kvs_count_val;
+            kvs_start = j + 2;
+            kvs_end = kvs_start + (kvs_count * 2);
+
+            /* Validate we have enough key-value pairs */
+            if (kvs_end > c->argc) {
+                addReplyError(c, "wrong number of key-value pairs");
+                return;
+            }
+            j = kvs_end - 1;  /* Skip directly to end of key-value block */
+        } else if (!strcasecmp(arg, "NX") && !(flags & OBJ_SET_XX)) {
+            flags |= OBJ_SET_NX;
+        } else if (!strcasecmp(arg, "XX") && !(flags & OBJ_SET_NX)) {
+            flags |= OBJ_SET_XX;
+        } else if (!strcasecmp(arg, "EX") && j + 1 < c->argc &&
+                   !(flags & (OBJ_PX | OBJ_EXAT | OBJ_PXAT | OBJ_KEEPTTL))) {
+            expire_flag_pos = j;  /* Remember position of EX flag */
```

Comment:
The following several places are similar as well

```suggestion
        } else if (!strcasecmp(arg, "EX") && j + 1 < c->argc &&
                   !(flags & (OBJ_PX | OBJ_EXAT | OBJ_PXAT | OBJ_KEEPTTL)))
        {
            expire_flag_pos = j;  /* Remember position of EX flag */
```